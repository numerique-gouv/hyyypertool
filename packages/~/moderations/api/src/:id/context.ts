//

import { NotFoundError } from "@~/app.core/error";
import type { App_Context } from "@~/app.middleware/context";
import { urls } from "@~/app.urls";
import { schema, type MonComptePro_PgDatabase } from "@~/moncomptepro.database";
import type { GetFicheOrganizationByIdHandler } from "@~/organizations.lib/usecase";
import { type get_domain_count_dto } from "@~/organizations.repository/get_domain_count";
import { type get_organization_members_count_dto } from "@~/organizations.repository/get_organization_members_count";
import { and, eq } from "drizzle-orm";
import type { Env, InferRequestType } from "hono";
import { useRequestContext } from "hono/jsx-renderer";

//

export interface ModerationContext extends Env {
  Variables: {
    moderation: get_moderation_dto;
  };
}
export interface ContextVariablesType extends Env {
  Variables: {
    domain: string;
    moderation: get_moderation_dto;
    organization_member: get_organization_member_dto;
    organization_fiche: Awaited<ReturnType<GetFicheOrganizationByIdHandler>>;
    query_organization_members_count: Promise<get_organization_members_count_dto>;
    query_domain_count: Promise<get_domain_count_dto>;
  };
}
export type ContextType = App_Context & ContextVariablesType;

//

const $get = typeof urls.moderations[":id"].$get;

type PageInputType = {
  out: InferRequestType<typeof $get>;
};

export const usePageRequestContext = useRequestContext<
  ContextType,
  any,
  PageInputType
>;

//

export async function get_organization_member(
  { pg }: { pg: MonComptePro_PgDatabase },
  { user_id, organization_id }: { user_id: number; organization_id: number },
) {
  return pg.query.users_organizations.findFirst({
    columns: { is_external: true },
    where: and(
      eq(schema.users_organizations.user_id, user_id),
      eq(schema.users_organizations.organization_id, organization_id),
    ),
  });
}
export type get_organization_member_dto = Awaited<
  ReturnType<typeof get_organization_member>
>;

//

export async function get_moderation(
  { pg }: { pg: MonComptePro_PgDatabase },
  { moderation_id }: { moderation_id: number },
) {
  const moderation = await pg.query.moderations.findFirst({
    where: eq(schema.moderations.id, moderation_id),
    with: {
      organization: {
        columns: {
          cached_activite_principale: true,
          cached_adresse: true,
          cached_code_postal: true,
          cached_est_active: true,
          cached_etat_administratif: true,
          cached_libelle_activite_principale: true,
          cached_libelle_categorie_juridique: true,
          cached_libelle_tranche_effectif: true,
          cached_libelle: true,
          cached_nom_complet: true,
          cached_tranche_effectifs: true,
          created_at: true,
          id: true,
          siret: true,
          updated_at: true,
        },
      },
      user: {
        columns: {
          created_at: true,
          email: true,
          family_name: true,
          given_name: true,
          id: true,
          job: true,
          last_sign_in_at: true,
          phone_number: true,
          sign_in_count: true,
        },
      },
    },
  });

  if (!moderation) throw new NotFoundError("Moderation not found.");
  return moderation;
}
export type get_moderation_dto = Awaited<ReturnType<typeof get_moderation>>;
