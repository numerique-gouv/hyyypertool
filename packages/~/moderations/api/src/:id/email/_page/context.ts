//

import type { App_Context } from "@~/app.middleware/context";
import { urls } from "@~/app.urls";
import type { get_crisp_mail_dto } from "@~/crisp.lib";
import type { Crisp_Context } from "@~/crisp.middleware";
import { type get_zammad_mail_dto } from "@~/zammad.lib/get_zammad_mail";
import type { Env, InferRequestType } from "hono";
import { useRequestContext } from "hono/jsx-renderer";
import type { Simplify } from "hono/utils/types";

//

export interface ContextVariablesType extends Env {
  Variables: {
    MAX_ARTICLE_COUNT: number;
    //
    crisp:
      | undefined
      | {
          conversation: get_crisp_mail_dto["conversation"];
          messages: get_crisp_mail_dto["messages"];
          session_id: string;
          show_more: boolean;
          subject: string;
        };
    moderation: Simplify<
      Pick<Moderation, "ticket_id"> & { user: Pick<User, "email"> }
    >;
    zammad:
      | undefined
      | {
          articles: get_zammad_mail_dto;
          show_more: boolean;
          subject: string;
          ticket_id: string;
        };
  };
}
export type ContextType = App_Context & Crisp_Context & ContextVariablesType;

//

const $get = typeof urls.moderations[":id"].email.$get;
type PageInputType = {
  out: InferRequestType<typeof $get>;
};

export const usePageRequestContext = useRequestContext<
  ContextType,
  any,
  PageInputType
>;
