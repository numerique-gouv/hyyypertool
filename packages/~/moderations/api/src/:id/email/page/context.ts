//

import { NotFoundError } from "@~/app.core/error";
import { DescribedBy_Schema, Entity_Schema } from "@~/app.core/schema";
import type { App_Context } from "@~/app.middleware/context";
import type { Crisp_Context } from "@~/crisp.middleware";
import type { GetCripsFromSessionIdHandler } from "@~/moderations.lib/usecase/GetCripsFromSessionId";
import { schema, type MonComptePro_PgDatabase } from "@~/moncomptepro.database";
import { type get_zammad_mail_dto } from "@~/zammad.lib/get_zammad_mail";
import { eq } from "drizzle-orm";
import { type Env } from "hono";
import { useRequestContext } from "hono/jsx-renderer";
import type { z } from "zod";

//

export interface ContextVariablesType extends Env {
  Variables: {
    MAX_ARTICLE_COUNT: number;
    //
    crisp?: Awaited<ReturnType<GetCripsFromSessionIdHandler>>;
    moderation: get_moderation_dto;
    zammad:
      | undefined
      | {
          articles: get_zammad_mail_dto;
          show_more: boolean;
          subject: string;
          ticket_id: string;
        };
  };
}
export type ContextType = App_Context & Crisp_Context & ContextVariablesType;

//

type PageInputType = {
  out: {
    param: z.input<typeof Entity_Schema>;
    query: z.input<typeof DescribedBy_Schema>;
  };
};

export const usePageRequestContext = useRequestContext<
  ContextType,
  any,
  PageInputType
>;

//

export async function get_moderation(
  pg: MonComptePro_PgDatabase,
  { moderation_id }: { moderation_id: number },
) {
  const moderation = await pg.query.moderations.findFirst({
    columns: { ticket_id: true },
    where: eq(schema.moderations.id, moderation_id),
    with: { user: { columns: { email: true } } },
  });

  if (!moderation) {
    throw new NotFoundError("Moderation not found");
  }

  return moderation;
}

export type get_moderation_dto = Awaited<ReturnType<typeof get_moderation>>;
