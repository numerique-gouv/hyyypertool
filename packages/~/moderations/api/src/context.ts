//

import { type Pagination } from "@~/app.core/schema";
import { z_coerce_boolean } from "@~/app.core/schema/z_coerce_boolean";
import { z_empty_string_to_undefined } from "@~/app.core/schema/z_empty_string_to_undefined";
import type { App_Context } from "@~/app.middleware/context";
import type { GetModerationsList } from "@~/moderations.core/queries/get_moderations_list/handler";
import type { Env } from "hono";
import { useRequestContext } from "hono/jsx-renderer";
import { z } from "zod";

//

export const MODERATION_TABLE_ID = "moderation_table";
export const MODERATION_TABLE_PAGE_ID = "moderation_table_page";

export const z_query = z.object({
  day: z
    .string()
    .default("")
    .pipe(z.coerce.date().or(z_empty_string_to_undefined)),
  siret: z.string().default(""),
  email: z.string().default(""),
  processed_requests: z.string().pipe(z_coerce_boolean).default("false"),
  hide_non_verified_domain: z.string().pipe(z_coerce_boolean).default("false"),
  hide_join_organization: z.string().pipe(z_coerce_boolean).default("false"),
});

//

interface GetModerationsListComand {
  get_moderations_list: GetModerationsList;
}
export type Commands = GetModerationsListComand;

//

export interface ViewModel {
  pagination: Pagination;
  search: z.infer<typeof z_query>;
  query_moderations_list: ReturnType<GetModerationsList>;
}

export interface ContextVariablesType extends Env {
  Variables: Commands & ViewModel;
}

export type ContextType = App_Context & ContextVariablesType;

//

export const usePageRequestContext = useRequestContext<ContextType>;
