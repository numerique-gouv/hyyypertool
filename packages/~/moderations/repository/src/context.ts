//

import type { MonComptePro_PgDatabase } from "@~/moncomptepro.database";

//

export interface MonComptePro_PostgresContext {
  pg: MonComptePro_PgDatabase;
}
