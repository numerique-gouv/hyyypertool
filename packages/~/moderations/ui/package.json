{
  "name": "@~/moderations.ui",
  "version": "1.0.0",
  "private": true,
  "type": "module",
  "imports": {
    "#ui/*": {
      "types": "./src/*/index.ts",
      "default": "./src/*/index.ts"
    }
  },
  "exports": {
    "./*": {
      "types": "./src/*/index.ts",
      "default": "./src/*/index.ts"
    }
  },
  "dependencies": {
    "@~/app.ui": "workspace:*",
    "@~/app.urls": "workspace:*",
    "@~/organizations.lib": "workspace:*",
    "@~/users.lib": "workspace:*",
    "@~/moderations.lib": "workspace:*",
    "hono": "4.6.9",
    "tailwind-variants": "0.2.1",
    "ts-pattern": "5.4.0"
  },
  "devDependencies": {
    "@~/crisp.lib": "workspace:*",
    "@~/config.typescript": "workspace:*"
  }
}
