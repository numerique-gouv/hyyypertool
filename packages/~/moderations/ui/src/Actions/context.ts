//

import { hyper_ref } from "@~/app.core/html";
import type { SimplifyDeep } from "@~/app.core/types";
import type { Moderation } from "@~/moderations.lib/entities/Moderation";
import type { IsUserExternalMemberHandler } from "@~/moderations.lib/usecase/IsUserExternalMember";
import type { Organization } from "@~/organizations.lib/entities/Organization";
import type { SuggestOrganizationDomainsHandler } from "@~/organizations.lib/usecase";
import type { User } from "@~/users.lib/entities/User";
import type { SuggestSameUserEmailsHandler } from "@~/users.lib/usecase/SuggestSameUserEmails";
import { createContext } from "hono/jsx";

export interface Values {
  domain: string;
  moderation: SimplifyDeep<
    Pick<Moderation, "id" | "moderated_at" | "type"> & {
      organization: Pick<
        Organization,
        "cached_libelle" | "id" | "siret" | "cached_libelle_categorie_juridique"
      >;
      user: Pick<User, "email" | "id" | "family_name" | "given_name">;
    }
  >;
  $reject: string;
  $accept: string;
  $decision_form: string;
  query_suggest_same_user_emails: SuggestSameUserEmailsHandler;
  query_is_user_external_member: IsUserExternalMemberHandler;
  query_suggest_organization_domains: SuggestOrganizationDomainsHandler;
}
export const context = createContext<Values>(null as any);

//
export const reject_context = createContext({
  $destination: hyper_ref(),
  $message: hyper_ref(),
  $object: hyper_ref(),
  $modal_message: hyper_ref(),
});
export const valid_context = createContext({
  $add_as_external_member: hyper_ref(),
  $add_as_internal_member: hyper_ref(),
  $add_domain: hyper_ref(),
  $send_notification: hyper_ref(),
  is_already_internal_member: false,
  is_already_external_member: false,
});
