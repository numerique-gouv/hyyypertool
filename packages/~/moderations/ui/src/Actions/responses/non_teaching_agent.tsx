import { useContext } from "hono/jsx";
import { dedent } from "ts-dedent";
import { context } from "../context";

export const label =
  "Agent hors enseignant : e-mail académique —> Min Edu Nat et Jeunesse";

export default function template() {
  const {
    moderation: {
      organization: { cached_libelle: organization_name },
      user: { email },
    },
  } = useContext(context);

  return dedent`
    Bonjour,

    Nous avons bien reçu votre demande de rattachement à l'organisation « ${organization_name} » sur ProConnect (anciennement : AgentConnect, MonComptePro).

    Votre adresse e-mail académique : « ${email} » ne vous permet pas de rattacher votre compte utilisateur MonComptePro au Ministère de l’Éducation Nationale et de la Jeunesse.

    En fonction de votre profession et de votre périmètre d’intervention, merci de bien vouloir créer votre compte à nouveau en le rattachant à l’une des organisations suivantes :

    - L’établissement scolaire dans lequel vous exercez,
    - La Direction des services départementaux de l'Éducation nationale de laquelle vous dépendez,
    - Le rectorat de votre Académie.

    Bien cordialement,
    L’équipe ProConnect.
  `;
}
