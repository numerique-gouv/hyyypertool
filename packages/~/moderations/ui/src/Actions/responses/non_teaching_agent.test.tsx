//

import { render_md } from "@~/app.ui/testing";
import { expect, test } from "bun:test";
import { context, type Values } from "../context";
import non_teaching_agent from "./non_teaching_agent";

//

test("returns missing name message", async () => {
  expect(
    await render_md(
      <context.Provider
        value={
          {
            moderation: {
              organization: { cached_libelle: "ðŸ¦„", siret: "ðŸ––" },
            },
          } as Values
        }
      >
        <Response />
      </context.Provider>,
    ),
  ).toMatchSnapshot();
});

function Response() {
  return <>{non_teaching_agent()}</>;
}
