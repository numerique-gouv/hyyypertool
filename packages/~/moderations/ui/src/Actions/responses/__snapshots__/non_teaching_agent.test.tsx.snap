// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`returns missing name message 1`] = `""`;
